### PR TITLE
Mm/disable tracking

### PR DESCRIPTION
**Pull request by MarcMcIntosh**
Allows for disabling tracking by setting `tracking: false` in sm.json

note:
when https://github.com/prismicio/slice-machine/pull/281/ is merged I'll update the configure-project test

## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] All new and existing tests are passing.

### CODE REVIEW

**Comment by MarcMcIntosh** [comment]
Interesting eslint plugin for next
```
The Next.js plugin was not detected in your ESLint configuration. See https://nextjs.org/docs/basic-features/eslint#migrating-existing-config 
```

File: packages/init/src/index.ts [hunk 1]
```diff
@@ -20,6 +20,8 @@ async function init() {
   const base = findArgument(process.argv, "base") || Utils.CONSTS.DEFAULT_BASE;
   const lib: string | undefined = findArgument(process.argv, "library");
   const branch: string | undefined = findArgument(process.argv, "branch");
+  const tracking =
```
Comment (Miniplop) [inline packages/init/src/index.ts]: There is no cleaner way to do this ?

Comment (Miniplop) [reply packages/init/src/index.ts]: Maybe we can rename the variable into "isTrackingAvailable" ?

File: packages/init/src/index.ts [hunk 2]
```diff
@@ -63,7 +70,7 @@ async function init() {
     : undefined;
 
   // configure the SM.json file and the json package file of the project..
-  configureProject(cwd, base, name, frameworkResult, sliceLibPath);
+  configureProject(cwd, base, name, frameworkResult, sliceLibPath, tracking);
```
Comment (Miniplop) [inline packages/init/src/index.ts]: You think that's necessary to write this information into the sm.json ? As we won't communicate about this option in the init.

File: packages/init/src/steps/configure-project.ts [hunk 3]
```diff
@@ -24,6 +25,7 @@ export function configureProject(
       apiEndpoint: Utils.Endpoints.buildRepositoryEndpoint(base, repository),
       libraries: ["@/slices", ...sliceLibPath],
       ...(framework.manuallyAdded ? { framework: framework.value } : {}),
+      ...(tracking === false ? { tracking } : {}),
```
Comment (Miniplop) [inline packages/init/src/steps/configure-project.ts]: can be simplified into `!tracking`

File: packages/init/src/utils/tracker.ts [hunk 4]
```diff
@@ -16,18 +16,20 @@ export class Tracker {
   constructor(
     readonly analytics: ServerAnalytics,
     readonly repo: string,
-    readonly identifier: { userId: string } | { anonymousId: string }
+    readonly identifier: { userId: string } | { anonymousId: string },
+    readonly tracking = true
```
Comment (Miniplop) [inline packages/init/src/utils/tracker.ts:20]: isTrackingDisable or isTrackignActive or something more explicit no ?

File: packages/init/src/utils/tracker.ts [hunk 5]
```diff
@@ -38,6 +40,7 @@ export class Tracker {
     eventType: EventType,
     attributes: Record<string, unknown> = {}
   ): void {
+    if (this.tracking === false) return void 0;
```
Comment (Miniplop) [inline packages/init/src/utils/tracker.ts]: using return alone here is not enough ?

File: packages/slice-machine/.eslintrc.json [hunk 6]
```diff
@@ -28,16 +28,10 @@
     "@typescript-eslint/ban-types": "warn",
     "@typescript-eslint/no-floating-promises": "warn",
     "@typescript-eslint/restrict-plus-operands": "warn",
-    "@typescript-eslint/no-inferrable-types": "warn",
     "@typescript-eslint/ban-ts-comment": "warn",
     "@typescript-eslint/unbound-method": "warn",
-    "@typescript-eslint/prefer-regexp-exec": "warn",
     "@typescript-eslint/no-empty-function": "warn",
-    "prefer-const": "warn",
```
Comment (Miniplop) [inline packages/slice-machine/.eslintrc.json:36]: cool

File: packages/slice-machine/server/src/api/http/common.ts [hunk 7]
```diff
@@ -29,11 +29,13 @@ export function WithEnv(
...
         "JfTfmHaATChc4xueS7RcCBsixI71dJIJ",
         env.repo,
-        identifier
+        identifier,
+        env.manifest.tracking
       );
 
     const reqWithEnv = (() => {
-      let r = req as any;
+      // TODO: this mutates req, so why not assign to req directly ?
```
Comment (Miniplop) [inline packages/slice-machine/server/src/api/http/common.ts:37]: I think this is a good question, I think this is only done for typing

File: packages/slice-machine/server/src/api/services/tracker.ts [hunk 8]
```diff
@@ -42,6 +46,7 @@ export class ServerTracker {
   }
 
   private groupEvent(traits: Record<string, unknown>): void {
+    if (this.tracking === false) return void 0;
```
Comment (Miniplop) [inline packages/slice-machine/server/src/api/services/tracker.ts]: return alone is sufficient no ?

File: packages/slice-machine/server/src/api/services/tracker.ts [hunk 9]
```diff
@@ -34,6 +36,8 @@ export class ServerTracker {
     eventType: EventType,
     attributes: Record<string, unknown> = {}
   ): void {
+    if (this.tracking === false) return void 0;
```
Comment (Miniplop) [inline packages/slice-machine/server/src/api/services/tracker.ts]: !tracking

File: packages/slice-machine/src/utils/tracker.ts [hunk 10]
```diff
@@ -38,6 +43,7 @@ export class ClientTracker {
     eventType: EventType,
     attributes: Record<string, unknown> = {}
   ): void {
+    if (this.tracking === false) return void 0;
```
Comment (Miniplop) [inline packages/slice-machine/src/utils/tracker.ts]: !tracking

Comment (Miniplop) [reply packages/slice-machine/src/utils/tracker.ts]: return alone also

File: packages/slice-machine/src/utils/tracker.ts [hunk 11]
```diff
@@ -46,6 +52,7 @@ export class ClientTracker {
   }
 
   private groupEvent(traits: Record<string, unknown>): void {
+    if (this.tracking === false) return void 0;
```
Comment (Miniplop) [inline packages/slice-machine/src/utils/tracker.ts]: !tracking + return alone

**Review by Miniplop — CHANGES_REQUESTED** [review]

File: packages/slice-machine/src/utils/tracker.ts (hunk 10 above)
Comment (MarcMcIntosh) [reply packages/slice-machine/src/utils/tracker.ts]: can do, there's a few extra warnings in this one :/

File: packages/slice-machine/src/utils/tracker.ts (hunk 11 above)
Comment (MarcMcIntosh) [reply packages/slice-machine/src/utils/tracker.ts]: I guess that should be fine :)

File: packages/init/src/steps/configure-project.ts (hunk 3 above)
Comment (MarcMcIntosh) [reply packages/init/src/steps/configure-project.ts]: I'll see what I can do, as the tracking can be undefined at this point

File: packages/init/src/index.ts (hunk 1 above)
Comment (MarcMcIntosh) [reply packages/init/src/index.ts]: as a cli argument, or as a variable?

File: packages/slice-machine/server/src/api/services/tracker.ts (hunk 8 above)
Comment (MarcMcIntosh) [reply packages/slice-machine/server/src/api/services/tracker.ts]: Yes, but I was also dealing with extra linter warnings.
I think I've found something better :)

File: packages/init/src/index.ts (hunk 2 above)
Comment (MarcMcIntosh) [reply packages/init/src/index.ts]: depends do we want to have to disable tracking in multiple places?

File: packages/init/src/steps/configure-project.ts (hunk 3 above)
Comment (Miniplop) [reply packages/init/src/steps/configure-project.ts]: how tracking can be undefined at this point as the param as a default value set to true ?

Comment (MarcMcIntosh) [reply packages/init/src/steps/configure-project.ts]: oh, I forgot about that

File: packages/init/src/index.ts (hunk 1 above)
Comment (Miniplop) [reply packages/init/src/index.ts]: just the variable :) I saw that you did it :)

**Comment by Miniplop** [comment]
LGTM

**Review by Miniplop — APPROVED** [review]

**Pull request merged**